File: Leetcode/sbgo/Leetcode706-设计哈希映射.go
package sbgo

import (
	"container/list"
)

const base = 769

type pp struct {
	key, value int
}

type MyHashMap struct {
	data []list.List
}

func (this *MyHashMap) hash(key int) int {
	return key % base
}

/** Initialize your data structure here. */
func Constructor() MyHashMap {
	return MyHashMap{make([]list.List, base)}
}

/** value will always be non-negative. */
func (this *MyHashMap) Put(key int, value int) {
	h := this.hash(key)
	for e := this.data[h].Front(); e != nil; e = e.Next() {
		if et := e.Value.(pp); et.key == key {
			e.Value = pp{key, value}
			return
		}
	}
	this.data[h].PushBack(pp{key, value})
}

/** Returns the value to which the specified key is mapped, or -1 if this map contains no mapping for the key */
func (this *MyHashMap) Get(key int) int {
	h := this.hash(key)
	for e := this.data[h].Front(); e != nil; e = e.Next() {
		if et := e.Value.(pp); et.key == key {
			return et.value
		}
	}
	return -1
}

/** Removes the mapping of the specified value key if this map contains a mapping for the key */
func (this *MyHashMap) Remove(key int) {
	h := this.hash(key)
	for e := this.data[h].Front(); e != nil; e = e.Next() {
		if e.Value.(pp).key == key {
			this.data[h].Remove(e)
			return
		}
	}
}

File: Leetcode/PY/Leetcode706-设计哈希映射.py
class MyHashMap:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.base = 769
        self.data = [[] for _ in range(self.base)]


    def hash(self,key):
        return key % self.base


    def put(self, key: int, value: int) -> None:
        """
        value will always be non-negative.
        """
        h = self.hash(key)
        for ii in self.data[h]:
            if ii[0] == key:
                ii[1] = value
                return
        self.data[h].append([key,value])


    def get(self, key: int) -> int:
        """
        Returns the value to which the specified key is mapped, or -1 if this map contains no mapping for the key
        """
        h = self.hash(key)
        for ii in self.data[h]:
            if ii[0] == key:
                return ii[1]
        return -1



    def remove(self, key: int) -> None:
        """
        Removes the mapping of the specified value key if this map contains a mapping for the key
        """
        h = self.hash(key)
        for i,ii in enumerate(self.data[h]):
            if ii[0] == key:
                self.data[h].pop(i)
                #py这里可以不return
                return 